pretyping/globEnv.mli

```diff
@@ -0,0 +1,83 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *   INRIA, CNRS and contributors - Copyright 1999-2018       *)
+(* <O___,, *       (see CREDITS file for the list of authors)           *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+open Names
+open Environ
+open Evd
+open EConstr
+open Ltac_pretype
+
+val empty_lvar : ltac_var_map
+
+(** To embed constr in glob_constr *)
+
+val register_constr_interp0 :
+  ('r, 'g, 't) Genarg.genarg_type ->
+    (unbound_ltac_var_map -> env -> evar_map -> types -> 'g -> constr * evar_map) -> unit
+
+(** {6 Pretyping name management} *)
+
+(** The following provides a level of abstraction for the kind of
+    environment used for type inference (so-called pretyping); in
+    particular:
+    - it supports that term variables can be interpreted as Ltac
+      variables pointing to the effective expected name
+    - it incrementally and lazily computes the renaming of rel
+      variables used to build purely-named evar contexts
+*)
+
+(** Type of environment extended with naming and ltac interpretation data *)
+
+type t
+
+(** Build a pretyping environment from an ltac environment *)
+
+val make : env -> evar_map -> ltac_var_map -> t
+
+(** Export the underlying environement *)
+
+val env : t -> env
+
+(** Push to the environment, returning the declaration(s) with interpreted names *)
+
+val push_rel : evar_map -> rel_declaration -> t -> rel_declaration * t
+val push_rel_context : ?force_names:bool -> evar_map -> rel_context -> t -> rel_context * t
+val push_rec_types : evar_map -> Name.t array * constr array -> t -> Name.t array * t
+
+(** Declare an evar using renaming information *)
+
+val e_new_evar : t -> evar_map ref -> ?src:Evar_kinds.t Loc.located ->
+  ?naming:Misctypes.intro_pattern_naming_expr -> constr -> constr
+
```

Comment:
I know that the pretyper is still using that style, but I'm not very fond of the imperative API.

Reply:
Yes, that was by consistency of style. Would you prefer e.g. that I move entirely `pretyping.ml` to a threaded `evd` (or `sigma`) style?

Reply:
I'm not pushing strongly for it, that was mostly a remark.

Reply:
I think it would be good to move to an explicit threading, yes. If not, I would prefer that we don't stay uniform rather than introducing more imperative interfaces.

Reply:
Note that I'd like that at some point we need to switch to the tactic API.

Reply:
OK, if noone has an opposite preference within the next hours, I shall make `e_new_evar` functional (and maybe with a different names to avoid the collision with the one if `evarutil.ml`).

Reply:
@herbelin note #6342